### PR TITLE
New: rest-spread-spacing rule (fixes #5391)

### LINKED ISSUE #5391: Rule Proposal: spread-spacing
I'd like to enforce spacing between the spread operator (`...`) and its expression.  I propose calling it `spread-spacing`

**When does this rule warn? Please describe and show example code:**

I envision it being configured with strings "always" and "never". 

With "never" (the default):

```js
// Warns
myFunction(... args);

// Does not warn
myFunction(...args);
```

And vice-versa for "always"

**Is this rule preventing an error or is it stylistic?**

Stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

It seems to fit with the other spacing rules we've included in core.

**Are you willing to create the rule yourself?**

Yes.

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

File: lib/rules/spread-spacing.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            alwaysSpace = context.options[0] === "always";
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        return {
+            SpreadElement: function(node) {
+                var args = node.argument,
+                    operator = sourceCode.getTokenBefore(args);
```
Comment (mysticatea) [inline lib/rules/spread-spacing.js]: I'd like to see a test for `...(foo)`.

Comment (kaicataldo) [inline lib/rules/spread-spacing.js]: Great catch, will update with the fix

**Comment by mysticatea** [comment]
As [this comment](https://github.com/eslint/eslint/issues/5391#issuecomment-188546399) mentioned, I think we should support rest elements and [experimentalObjectRestSpread](http://eslint.org/docs/user-guide/configuring#specifying-parser-options).

**Comment by kaicataldo** [comment]
Missed that. Thanks, will update and report back.

**Comment by kaicataldo** [comment]
Thanks for pointing that stuff out, @mysticatea :) Updated

File: lib/rules/rest-spread-spacing.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                    break;
+                default:
+                    return;
+            }
+
+            while (operator.value !== "...") {
+                operator = sourceCode.getTokenBefore(operator);
+            }
+
+            if (alwaysSpace && !sourceCode.isSpaceBetweenTokens(operator, args)) {
```
Comment (mysticatea) [inline lib/rules/rest-spread-spacing.js]: Maybe `"never"` option of this rule is conflicting with the `"always"` option of `space-in-parens` rule: e.g. `...( foo )`.
I think this should check spacing between `operator` and the next token of `operator`, as same as fix functions.

Comment (kaicataldo) [inline lib/rules/rest-spread-spacing.js]: Makes sense - updated once more!

**Comment by mysticatea** [comment]
Thank you for great works!
LGTM.

(Though I cannot check documents!)

File: tests/lib/rules/rest-spread-spacing.js [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{
+                line: 1,
+                column: 7,
+                message: "Unexpected whitespace after spread operator",
+                type: "SpreadElement"
+            }]
+        },
+        {
+            code: "fn(... ( args ))",
```
Comment (kaicataldo) [inline tests/lib/rules/rest-spread-spacing.js]: Added tests to ensure compatibility with `space-in-parens`

**Comment by kaicataldo** [comment]
@mysticatea Refactored and simplified [this logic](https://github.com/eslint/eslint/pull/6278/files#diff-52fdcb49902e6e148971fa73f2f9f069R41) after changing to account for parens. Thanks again for reviewing 👍

**Comment by nzakas** [comment]
Lgtm. Nice job!

**Pull request merged**